File: Graph_algo/Bellman_Ford/BellmanFord.cpp
#include <iostream>
#include <vector>
#include <limits.h>
using namespace std;

struct Edge {
    int u, v, weight;
};

class BellmanFord {
public:
    int V; // Number of vertices
    vector<Edge> edges; // List of edges

    BellmanFord(int vertices) {
        V = vertices;
    }

    void addEdge(int u, int v, int weight) {
        edges.push_back({u, v, weight});
    }

    void findShortestPath(int src) {
        // Step 1: Initialize distances from src to all other vertices as infinite
        vector<int> dist(V, INT_MAX);
        dist[src] = 0;

        // Step 2: Relax all edges |V| - 1 times
        for (int i = 1; i <= V - 1; i++) {
            for (auto edge : edges) {
                if (dist[edge.u] != INT_MAX && dist[edge.u] + edge.weight < dist[edge.v]) {
                    dist[edge.v] = dist[edge.u] + edge.weight;
                }
            }
        }

        // Step 3: Check for negative weight cycles
        for (auto edge : edges) {
            if (dist[edge.u] != INT_MAX && dist[edge.u] + edge.weight < dist[edge.v]) {
                cout << "Graph contains a negative weight cycle." << endl;
                return;
            }
        }

        // Print the shortest distances
        cout << "Shortest distances from source vertex " << src << ":" << endl;
        for (int i = 0; i < V; i++) {
            if (dist[i] == INT_MAX)
                cout << "Vertex " << i << " : Distance INF" << endl;
            else
                cout << "Vertex " << i << " : Distance " << dist[i] << endl;
        }
    }
};

int main() {
    // Create a graph with 5 vertices
    BellmanFord graph(5);

    // Add edges
    graph.addEdge(0, 1, -1);
    graph.addEdge(0, 2, 4);
    graph.addEdge(1, 2, 3);
    graph.addEdge(1, 3, 2);
    graph.addEdge(1, 4, 2);
    graph.addEdge(3, 2, 5);
    graph.addEdge(3, 1, 1);
    graph.addEdge(4, 3, -3);

    // Find the shortest paths from source vertex 0
    graph.findShortestPath(0);

    return 0;
}

File: Graph_algo/Bellman_Ford/BellmanFord.py
class BellmanFord:
    def __init__(self, vertices):
        self.V = vertices  # Number of vertices
        self.edges = []    # List to store all edges

    def add_edge(self, u, v, weight):
        # Add directed edge (u -> v) with a given weight
        self.edges.append((u, v, weight))

    def shortest_path(self, src):
        # Step 1: Initialize distances from src to all other vertices as infinite
        dist = [float('inf')] * self.V
        dist[src] = 0

        # Step 2: Relax all edges |V| - 1 times
        for _ in range(self.V - 1):
            for u, v, weight in self.edges:
                if dist[u] != float('inf') and dist[u] + weight < dist[v]:
                    dist[v] = dist[u] + weight

        # Step 3: Check for negative weight cycles
        for u, v, weight in self.edges:
            if dist[u] != float('inf') and dist[u] + weight < dist[v]:
                print("Graph contains a negative weight cycle.")
                return None

        # Return the computed shortest distances
        return dist


# Example Usage
if __name__ == "__main__":
    g = BellmanFord(5)
    g.add_edge(0, 1, -1)
    g.add_edge(0, 2, 4)
    g.add_edge(1, 2, 3)
    g.add_edge(1, 3, 2)
    g.add_edge(1, 4, 2)
    g.add_edge(3, 2, 5)
    g.add_edge(3, 1, 1)
    g.add_edge(4, 3, -3)

    source = 0
    distances = g.shortest_path(source)
    if distances:
        print("Shortest distances from source vertex", source, ":")
        for i, d in enumerate(distances):
            print(f"Vertex {i} : Distance {d}")
